File: ios-app/QingJiCore/Sources/QingJiCore/Budget/BudgetEngine.swift
import Foundation

/// 预算周期，与 Android 端的月度、周度和自定义期间保持同一语义。
public enum BudgetCycle: String, Codable, CaseIterable, Hashable, Sendable {
    case monthly
    case weekly
    case custom
}

/// 一个预算周期的左闭右开日期窗口。
public struct BudgetWindow: Equatable, Sendable {
    public let startInclusive: Date
    public let endExclusive: Date
    public let cycle: BudgetCycle

    public init(startInclusive: Date, endExclusive: Date, cycle: BudgetCycle) {
        self.startInclusive = startInclusive
        self.endExclusive = endExclusive
        self.cycle = cycle
    }
}

/// 当月预算执行状态。
public struct BudgetStatus: Equatable, Sendable {
    public let monthlyBudget: Decimal
    public let spentThisMonth: Decimal
    public let spentToday: Decimal
    /// 月剩余预算（可为负）。
    public let remaining: Decimal
    /// 「今日可花」：(预算 − 今天之前已花) ÷ 含今天的剩余天数 − 今天已花。可为负。
    public let todayAllowance: Decimal
    public let isOverBudget: Bool

    public init(
        monthlyBudget: Decimal,
        spentThisMonth: Decimal,
        spentToday: Decimal,
        remaining: Decimal,
        todayAllowance: Decimal,
        isOverBudget: Bool
    ) {
        self.monthlyBudget = monthlyBudget
        self.spentThisMonth = spentThisMonth
        self.spentToday = spentToday
        self.remaining = remaining
        self.todayAllowance = todayAllowance
        self.isOverBudget = isOverBudget
    }
}

/// 预算前置提醒的核心：把「月底才知道超支」变成「今天就知道还能花多少」。
public enum BudgetEngine {
    /// 根据参考日期构造周期窗口。自定义周期的结束日期按“当天结束”处理。
    public static func window(
        cycle: BudgetCycle,
        referenceDate: Date,
        customStart: Date? = nil,
        customEnd: Date? = nil,
        calendar: Calendar = .current
    ) -> BudgetWindow? {
        let day = calendar.startOfDay(for: referenceDate)
        switch cycle {
        case .monthly:
            let components = calendar.dateComponents([.year, .month], from: day)
            guard let start = calendar.date(from: DateComponents(
                year: components.year,
                month: components.month,
                day: 1
            )), let end = calendar.date(byAdding: .month, value: 1, to: start) else {
                return nil
            }
            return BudgetWindow(startInclusive: start, endExclusive: end, cycle: cycle)
        case .weekly:
            let weekday = calendar.component(.weekday, from: day)
            let mondayOffset = (weekday + 5) % 7
            guard let start = calendar.date(byAdding: .day, value: -mondayOffset, to: day),
                  let end = calendar.date(byAdding: .day, value: 7, to: start) else {
                return nil
            }
            return BudgetWindow(startInclusive: start, endExclusive: end, cycle: cycle)
        case .custom:
            guard let customStart, let customEnd else { return nil }
            let start = calendar.startOfDay(for: customStart)
            guard let inclusiveEnd = calendar.date(
                byAdding: .day,
                value: 1,
                to: calendar.startOfDay(for: customEnd)
            ), start < inclusiveEnd else { return nil }
            return BudgetWindow(startInclusive: start, endExclusive: inclusiveEnd, cycle: cycle)
        }
    }

    /// 任意预算窗口的执行状态；月度旧接口继续保留给旧调用方。
    public static func status(
        budget: Decimal,
        cycle: BudgetCycle,
        referenceDate: Date = Date(),
        customStart: Date? = nil,
        customEnd: Date? = nil,
        categoryKey: String? = nil,
        records: [TransactionRecord],
        calendar: Calendar = .current
    ) -> BudgetStatus {
        guard let window = window(
            cycle: cycle,
            referenceDate: referenceDate,
            customStart: customStart,
            customEnd: customEnd,
            calendar: calendar
        ) else {
            return BudgetStatus(
                monthlyBudget: budget,
                spentThisMonth: 0,
                spentToday: 0,
                remaining: budget,
                todayAllowance: budget,
                isOverBudget: false
            )
        }

        let day = calendar.startOfDay(for: referenceDate)
        let today = day < window.startInclusive
            ? window.startInclusive
            : (day >= window.endExclusive
                ? calendar.date(byAdding: .day, value: -1, to: window.endExclusive) ?? day
                : day)
        let userRecords = LedgerPolicy.userRecords(from: records)
        var spent = Decimal.zero
        var spentToday = Decimal.zero
        for record in userRecords where record.kind == .expense && record.amount > 0 {
            if let categoryKey,
               record.categoryKey != categoryKey && record.topCategoryKey != categoryKey {
                continue
            }
            let recordDay = calendar.startOfDay(for: record.date)
            guard recordDay >= window.startInclusive, recordDay < window.endExclusive else {
                continue
            }
            spent += record.amount
            if recordDay == today { spentToday += record.amount }
        }

        let spentBeforeToday = spent - spentToday
        let remainingDays = max(
            calendar.dateComponents([.day], from: today, to: window.endExclusive).day ?? 1,
            1
        )
        let todayAllowance = (budget - spentBeforeToday) / Decimal(remainingDays) - spentToday
        return BudgetStatus(
            monthlyBudget: budget,
            spentThisMonth: spent,
            spentToday: spentToday,
            remaining: budget - spent,
            todayAllowance: todayAllowance,
            isOverBudget: spent > budget
        )
    }

    public static func status(
        monthlyBudget: Decimal,
        records: [TransactionRecord],
        on date: Date = Date(),
        calendar: Calendar = .current
    ) -> BudgetStatus {
        let components = calendar.dateComponents([.year, .month, .day], from: date)
        let today = components.day ?? 1
        let dayCount = StatisticsEngine.daysInMonth(
            year: components.year ?? 2000,
            month: components.month ?? 1,
            calendar: calendar
        )

        let userRecords = LedgerPolicy.userRecords(from: records)
        var spentThisMonth = Decimal(0)
        var spentToday = Decimal(0)
        for record in userRecords where record.kind == .expense {
            let recordComponents = calendar.dateComponents([.year, .month, .day], from: record.date)
            guard recordComponents.year == components.year,
                  recordComponents.month == components.month else { continue }
            guard record.amount > 0 else { continue }
            spentThisMonth += record.amount
            if recordComponents.day == today {
                spentToday += record.amount
            }
        }

        let spentBeforeToday = spentThisMonth - spentToday
        let remainingDays = Decimal(max(dayCount - today + 1, 1))
        let todayAllowance = (monthlyBudget - spentBeforeToday) / remainingDays - spentToday

        return BudgetStatus(
            monthlyBudget: monthlyBudget,
            spentThisMonth: spentThisMonth,
            spentToday: spentToday,
            remaining: monthlyBudget - spentThisMonth,
            todayAllowance: todayAllowance,
            isOverBudget: spentThisMonth > monthlyBudget
        )
    }
}

File: android-app/lib/core/budget/budget_engine.dart
import 'package:decimal/decimal.dart';
import 'budget_window_resolver.dart';
import '../models/transaction_kind.dart';
import '../models/transaction_record.dart';
import '../statistics/statistics_engine.dart';

/// 当月预算执行状态。
class BudgetStatus {
  final Decimal monthlyBudget;
  final Decimal spentThisMonth;
  final Decimal spentToday;

  /// 月剩余预算（可为负）。
  final Decimal remaining;

  /// 「今日可花」：(预算 − 今天之前已花) ÷ 含今天的剩余天数 − 今天已花。可为负。
  ///
  /// 口径说明（和预算页「往后每天可花」区分，两者同一基底不矛盾）：
  ///   今日可花 = 今天这一天的份额，扣掉今天已花；
  ///   往后每天可花 = 剩余额度 ÷ 剩余天数（预算页自算），是往后的平均。
  final Decimal todayAllowance;
  final bool isOverBudget;

  const BudgetStatus({
    required this.monthlyBudget,
    required this.spentThisMonth,
    required this.spentToday,
    required this.remaining,
    required this.todayAllowance,
    required this.isOverBudget,
  });

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is BudgetStatus &&
          monthlyBudget == other.monthlyBudget &&
          spentThisMonth == other.spentThisMonth &&
          spentToday == other.spentToday &&
          remaining == other.remaining &&
          todayAllowance == other.todayAllowance &&
          isOverBudget == other.isOverBudget;

  @override
  int get hashCode => Object.hash(
        monthlyBudget,
        spentThisMonth,
        spentToday,
        remaining,
        todayAllowance,
        isOverBudget,
      );
}

/// 预算前置提醒的核心：把「月底才知道超支」变成「今天就知道还能花多少」。
class BudgetEngine {
  BudgetEngine._();

  /// Adapts the unified resolver output for legacy consumers.
  ///
  /// No configured plan (or an unavailable/conflicting spend result) remains
  /// null instead of becoming a zero budget. Historical windows have no
  /// current-cycle daily guidance, so their two legacy daily fields are zero.
  static BudgetStatus? fromWindowResult(BudgetWindowResult result) {
    final budget = result.plannedAmount;
    final spent = result.spentAmount;
    final remaining = result.remainingAmount;
    if (budget == null || spent == null || remaining == null) return null;

    final daily = result.currentCycleDailyStatus;
    return BudgetStatus(
      monthlyBudget: budget,
      spentThisMonth: spent,
      spentToday: daily?.spentTodayAmount ?? Decimal.zero,
      remaining: remaining,
      todayAllowance: daily?.todayRemainingAllowanceAmount ?? Decimal.zero,
      isOverBudget: remaining < Decimal.zero,
    );
  }

  static BudgetStatus status({
    required Decimal monthlyBudget,
    required List<TransactionRecord> records,
    DateTime? on,
  }) {
    final date = on ?? DateTime.now();
    final today = date.day;
    final dayCount = StatisticsEngine.daysInMonth(
      year: date.year,
      month: date.month,
    );

    var spentThisMonth = Decimal.zero;
    var spentToday = Decimal.zero;

    for (final record in records) {
      if (record.kind != TransactionKind.expense) continue;
      if (record.date.year != date.year) continue;
      if (record.date.month != date.month) continue;
      spentThisMonth += record.amount;
      if (record.date.day == today) {
        spentToday += record.amount;
      }
    }

    final spentBeforeToday = spentThisMonth - spentToday;
    final remainingDaysInt =
        dayCount - today + 1 < 1 ? 1 : dayCount - today + 1;
    final remainingDays = Decimal.fromInt(remainingDaysInt);
    // In decimal v3, Decimal / Decimal returns Rational; convert back to Decimal
    // with sufficient scale (10 decimal places) before subtracting spentToday.
    final quotient = ((monthlyBudget - spentBeforeToday) / remainingDays)
        .toDecimal(scaleOnInfinitePrecision: 10);
    final todayAllowance = quotient - spentToday;

    return BudgetStatus(
      monthlyBudget: monthlyBudget,
      spentThisMonth: spentThisMonth,
      spentToday: spentToday,
      remaining: monthlyBudget - spentThisMonth,
      todayAllowance: todayAllowance,
      isOverBudget: spentThisMonth > monthlyBudget,
    );
  }
}
